glommio/src/sys/source.rs

```diff
@@ -225,10 +226,17 @@ impl Source {
     // adds a single waiter to the list, replacing any waiter that may already
     // exist. Should be used for single streams that map a future 1:1 to their I/O
     // source
-    pub(crate) fn add_waiter_single(&self, waker: Waker) {
+    pub(crate) fn add_waiter_single(&self, waker: &Waker) {
         let mut inner = self.inner.borrow_mut();
-        inner.wakers.waiters.pop();
-        inner.wakers.waiters.push(waker);
+        let waiters = &mut inner.wakers.waiters;
+        match waiters.first_mut() {
```

Comment:
Can you explain the purpose behind this ?

Reply:
From what I can see, this is meant to add the `Waker` if it's different than the one already in place. 
This looks wrong because `Waker::will_wake` makes zero guarantees. i.e., it could return `true` when two wakers would wake different tasks.

Maybe the usage is safe, and this is sound, but a comment would be great so we can understand what you're trying to do. This code is susceptible, and we introduced some pretty nasty bugs here before :grimacing:

Reply:
While I was trying to understand the difference between glommio and tokio's socket implementation, I found that tokio used the same trick to prevent avoid cloning a waker. So I modified `add_waiter_single` in the same way to exclude noise in the benchmark. It turns out it made little difference whether we have this change or not. I keep it however, because it is logically very likely the existing waker and the one being added will wake the same task.